src/partial_quicksync.rs

```diff
@@ -0,0 +1,265 @@
+    let to: i64 = items[1].parse().unwrap();
+    if to > layer_from {
+      target_index = index;
+      break;
+    }
+  }
+  target_index = target_index - jump_back;
+  for line in metadata.lines().skip(target_index) {
+    result.push(RestorePoint::from_str(line).unwrap());
+  }
+  result
+}
+
+fn get_latest_from_db(target_path: &str) -> Result<i64> {
+  let conn = Connection::open(target_path)?;
+  conn
+    .query_row(
+      "SELECT max(id) FROM layers WHERE applied_block IS NOT null",
+      [],
+      |row| row.get(0),
+    )
+    .context("Failed to get latest layer from DB")
+}
+
+fn get_user_version(target_path: &str) -> Result<i64> {
+  let conn = Connection::open(target_path)?;
+  conn
+    .query_row("PRAGMA user_version", [], |row| row.get(0))
+    .context("Failed to get user version")
+}
+
+fn download_file(
+  client: &Client,
+  user_version: i64,
+  layer_from: i64,
+  layer_to: i64,
+  hash: &str,
+  target_path: &str,
+) -> Result<()> {
+  let base_url = get_base_url();
+  let suffix = if target_path.ends_with("zst") {
+    ".zst"
+  } else {
+    ""
+  };
+  let url = format!(
+    "{}/{}/{}_{}_{}/state.sql_diff.{}_{}.sql{}",
+    base_url, user_version, layer_from, layer_to, hash, layer_from, layer_to, suffix
+  );
+  println!("Downloading from {}", url);
+  let mut resp = client.get(&url).send().context("Failed to send request")?;
+  if !resp.status().is_success() {
+    anyhow::bail!(
+      "Failed to download file {}: HTTP status {}",
+      url,
+      resp.status()
+    );
+  }
+  let mut file = File::create(target_path).context("Failed to create file")?;
+  resp
+    .copy_to(&mut file)
+    .context("Failed to copy response to file")?;
+  Ok(())
+}
+
+fn decompress_file(input_path: &str, output_path: &str) -> Result<()> {
+  let input_file = File::open(input_path).context("Failed to open input file")?;
+  let output_file = File::create(output_path).context("Failed to create output file")?;
+
+  let mut reader = BufReader::new(input_file);
+  let mut writer = BufWriter::new(output_file);
+
+  let mut decoder = Decoder::new(&mut reader).context("Failed to create decoder")?;
+  decoder
+    .window_log_max(31)
+    .context("Failed to set window log max")?;
+
+  io::copy(&mut decoder, &mut writer).context("Failed to decompress")?;
+
+  Ok(())
+}
+
+fn execute_restore(target_conn: &Connection, restore_string: &str) -> Result<()> {
+  target_conn
+    .execute_batch(restore_string)
+    .context("Failed to execute restore")
+}
+
+pub fn partial_restore(layer_from: i64, target_db_path: &str, jump_back: usize) -> Result<()> {
+  let client = Client::new();
+  let base_url = get_base_url();
+  let user_version = get_user_version(target_db_path)?;
+  let remote_metadata = client
+    .get(format!("{}/{}/metadata.csv", base_url, user_version))
+    .send()?
+    .text()?;
+  let restore_string = client
+    .get(format!("{}/{}/restore.sql", base_url, user_version))
+    .send()?
+    .text()?;
```

Comment:
Why keep the restore SQL query on the server instead of in the code?

Reply:
To support multiple versions of the schema in target. Othewise we would need to hardcode these for every USER_VERSION.